CMakeLists.txt

```diff
@@ -480,18 +480,21 @@ if(ENABLE_HARDENING)
     try_append_linker_flag("/HIGHENTROPYVA" TARGET hardening_interface)
     try_append_linker_flag("/NXCOMPAT" TARGET hardening_interface)
   else()
+
+    # _FORTIFY_SOURCE requires that there is some level of optimization,
+    # otherwise it does nothing and just creates a compiler warning.
     try_append_cxx_flags("-U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=3"
       RESULT_VAR cxx_supports_fortify_source
+      SOURCE "int main() {
+              # if !defined __OPTIMIZE__ || __OPTIMIZE__ <= 0
```

Comment:
Could you provide a few links that confirm this code checks the source fortification feature?

Reply:
This just checks that optimisations are in use, which is when we want to use it, using the same approach as glibc: https://github.com/bminor/glibc/blob/4945ffc88a8ad49280bae64165683ddfd12b2390/include/features.h#L421.

Reply:
i.e The new code should be performing the same tests as master, except that, rather than outsourcing the check for optimisations to a test of whether or not the build mode is `Debug`, it tests if optimisations are being used directly.

Reply:
The check above reminds me of the part of the XZ backdoor where the attacker inserted a single . character to disable a linux security feature in a similar cmake check:

https://git.tukaani.org/?p=xz.git;a=blobdiff;f=CMakeLists.txt;h=d2b1af7ab0ab759b6805ced3dff2555e2a4b3f8e;hp=76700591059711e3a4da5b45cf58474dac4e12a7;hb=328c52da8a2bbb81307644efdb58db2c422d9ba7;hpb=eb8ad59e9bab32a8d655796afd39597ea6dcc64d

So I think it might be useful to add a sanity check to confirm that `cxx_supports_fortify_source` is actually true in release builds:

```cmake
if(NOT cxx_supports_fortify_source AND CMAKE_BUILD_TYPE MATCHES "Release")
    message(FATAL_ERROR "Error: Current release build flags are incompatible with fortify_source")
endif()
```

Separately, one theoretical drawback of this new implementation is it probably disables fortify source unconditionally in multi-configuration cmake backends like visual studio and xcode, whereas the previous `$<$<NOT:$<CONFIG:Debug>>:` generator expression was more flexible and it could enable fortify source in release configurations while disabling it in debug configurations. But I think this doesn't matter in practice because source fortification is a GNU feature which presumably doesn't work in visual studio or xcode anyway.

Reply:
I am certainly open to additional approaches in regards to the (potential fragility) of online source.

> So I think it might be useful to add a sanity check to confirm that cxx_supports_fortify_source is actually true in release builds:

I have a PR to try and do this in #27038.

Reply:
> I have a PR to try and do this in #27038.

Nice that seems potentially more reliable than suggested sanity check

Reply:
Can you elaborate, the issue is that one of the warnings is something we don't want to turn into an error, otherwise the check will fail incorrectly?

Reply:
Yea, I'm not sure that making this logic more complicated is worth it, to suppress (harmless) warnings, given that hardening also continues to work for GCC 11 & FORTIFY=3 (2).